core: using native interrupt handlers

etienne-lms opened the pull request:
The P-R is related to https://github.com/OP-TEE/optee_os/pull/5954.

It adds a section in core architecture documentation on how to implement and register interrupt controllers and handler functions.

jforissier commented on architecture/core.rst:
```diff
@@ -390,6 +390,183 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+Interrupt controllers are represented by an instance of ``struct itr_chip``.
+An interrupt controller exposes a given number of interrupts, each identified
+by an index from 0 to N-1 where N is the total number of interrupts exposed
+by that controller. In the literature, an interrupt index identifier
```
litterature

jforissier commented on architecture/core.rst:
```diff
@@ -390,6 +390,183 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+**Interrupt management API functions**
+
+Interrupt management resources are declared in header file
+kernel/interrupt.h_. Main API function are:
```
``` ``<kernel/interrupt.h>`` ``` maybe?

jforissier replied:
Yes fine with me.

jforissier commented on architecture/core.rst:
```diff
@@ -390,6 +390,183 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+kernel/interrupt.h_. Main API function are:
+
+    - ``interrupt_enable()`` and ``interrupt_disable()`` to respectively
+      enable and disable and interrupt.
```
an

jforissier commented on architecture/core.rst:
```diff
@@ -390,6 +390,183 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+      enable and disable and interrupt.
+
+    - ``interrupt_mask()`` and ``interrupt_unmask()`` to respectively
+      mask/unmask and interrupt.
```
an

jforissier commented on architecture/core.rst:
```diff
@@ -390,6 +390,183 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+When configuration switch ``CFG_DRIVERS_INTERRUPT`` is enabled, core
+offers a means to use Device Tree to describe interrupts topologyin the
+platform. When used, the platform must provide a secure devicetree blob
+(DTB) which platform and device drivers can use to getreferences to
```
get references

jforissier commented on architecture/core.rst:
```diff
@@ -390,6 +390,183 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+offers a means to use Device Tree to describe interrupts topologyin the
+platform. When used, the platform must provide a secure devicetree blob
+(DTB) which platform and device drivers can use to getreferences to
+interrupt controllers and relative interrupt numbers theyconsume.
```
they consume

jforissier commented on architecture/core.rst:
```diff
@@ -390,6 +390,183 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+platform. When used, the platform must provide a secure devicetree blob
+(DTB) which platform and device drivers can use to getreferences to
+interrupt controllers and relative interrupt numbers theyconsume.
+Main API functions are:
```
The main (some definite articles are missing in a few other places I believe but my English is not good enough to decide it omitting them is OK or not).

jforissier commented on architecture/core.rst:
```diff
@@ -390,6 +390,183 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+Main API functions are:
+
+    - ``dt_register_interrupt_provider()`` to register an interrupt
+      controller related to a given node in the Device Tree. This function
```
Shouldn't the function be called `dt_register_interrupt_controller()` then?

etienne-lms replied:
I understand your point of view. However all functions used the register a provider of device driver reference in _dt_drvier.c_ (and friends) resources are labelled `dt_register_xxx_provider()` so I would prefer to keep that `_provider` suffix. Here I could propose `dt_register_interrupt_controller_provider()` but it's a bit long, almost as long  as existing `dt_driver_device_from_node_idx_prop_phandle()` which I feel pretty bad about :(

What about `dt_register_itr_chip_provider()`?

jforissier replied:
Ah, consistency again ;) it's fine to keep `dt_register_interrupt_provider()` then.

jforissier commented on architecture/core.rst:
```diff
@@ -390,6 +390,183 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+function handlers for management of the interrupt. An interrupt controller
+must provide operation handler functions ``add``, ``enable`` and ``disable``.
+There are other operation handler functions but these are not mandatory, as
+``mask``, ``unmask``, ``rasie_pi``, ``raise_sgi`` and ``set_priority``.
```
raise
Open question: should parentheses be used when documenting function pointers?

jforissier commented on architecture/core.rst:
```diff
@@ -390,6 +390,183 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+**Interrrupt handlers**
+
+Interrupt handler functions are functions registered by drivers that
+core shall call when the related interrupt occurs. Structure struct itr_handler
```
``` ``struct itr_handler`` ```

jforissier commented on architecture/core.rst:
```diff
@@ -390,6 +390,183 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+core shall call when the related interrupt occurs. Structure struct itr_handler
+references a handler. It contains the handler function entry point as well
+as interrupt information: its number, its interrupt controller device and
+a few fields.
```
a few other fields

jforissier commented on architecture/core.rst:
```diff
@@ -390,6 +390,183 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+``ITRR_NONE`` when the interrupt cannot be served.
+
+The interrupt handler executes in an interrupt context, not in a thread
+context. The handler must therefore execute very quickly. If there are
```
"as quickly as possible"?
Maybe say why it is so (what happens when a handler runs for a long time)

jforissier commented on architecture/core.rst:
```diff
@@ -390,6 +390,183 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+
+**Using device tree**
+
+Device drivers can rely on Device Tree technology to get references to the
```
I would remove "technology"

etienne-lms replied:
ok

jforissier reviewed (commented):
Comments below, then:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented:
LGTM you may squash the commits if you want, I will hold off merging this until there is code to actually implement it.

jbech-linaro commented on architecture/core.rst:
```diff
@@ -390,6 +390,186 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+      respectively register and unregister an interrupt handler function
+      for a given interrupt.
+
+When configuration switch ``CFG_DRIVERS_INTERRUPT`` is enabled, core
+offers a means to use Device Tree to describe interrupts topologyin the
+platform. When used, the platform must provide a secure devicetree blob
```
There is a spelling error "topologyin", but I suggest rewriting the sentence slightly.
```suggestion
When the configuration switch ``CFG_DRIVERS_INTERRUPT`` is enabled,
the core provides a way to describe the platform's interrupt topology
using Device Tree.
```

jbech-linaro replied:
The "When used, ..." should be there. It's just the one sentence I had in mind.

etienne-lms replied:
Thanks. I will also replace `CFG_DRIVERS_INTERRUPT` with `CFG_DT`. `CFG_DRIVERS_INTERRUPT` is useless.

jbech-linaro commented on architecture/core.rst:
```diff
@@ -390,6 +390,186 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+Core handlers for native interrupts
+===================================
+
+OP-TEE core offers means to implement native interrupt
+(see :ref:`native_foreign_irqs`) controller drivers and means for device
+drivers to configure and register handler functions for these interrupts.
```
```suggestion
OP-TEE core provides methods for device drivers to setup and register
handler functions for native interrupt controller drivers
(see:ref:`native_foreign_irqs`).
```

jbech-linaro commented on architecture/core.rst:
```diff
@@ -390,6 +390,186 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+Interrupt controllers are represented by an instance of ``struct itr_chip``.
+An interrupt controller exposes a given number of interrupts, each identified
+by an index from 0 to N-1 where N is the total number of interrupts exposed
+by that controller. In the litterature, an interrupt index identifier
```
s/litterature/literature/

etienne-lms replied:
@jforissier and I should take my English lessons :)

jbech-linaro replied:
To be fair, I'm just as much native English speaker as you are 😄

etienne-lms replied:
But you're still a good reference for me. :beers:

jbech-linaro commented on architecture/core.rst:
```diff
@@ -390,6 +390,186 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+
+**Interrupt controller drivers**
+
+An interrupt controller instance (``struct itr_chip``) registers operation
+function handlers for management of the interrupt it controls. An interrupt
+controller
+must provide operation handler functions ``.add``, ``.enable`` and ``.disable``.
+There are other operation handler functions but these are not mandatory, as
+``.mask``, ``.unmask``, ``.rasie_pi``, ``.raise_sgi`` and ``.set_priority``.
```
Text not changed in the suggestion below, it's just that the word "controller" was alone on an entire line.

```suggestion
An interrupt controller instance (``struct itr_chip``) registers operation
function handlers for management of the interrupt it controls. An interrupt
controller must provide operation handler functions ``.add``, ``.enable`` and
``.disable``. There are other operation handler functions but these are not
mandatory, as ``.mask``, ``.unmask``, ``.rasie_pi``, ``.raise_sgi`` and
``.set_priority``.
```

jbech-linaro commented on architecture/core.rst:
```diff
@@ -390,6 +390,186 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+Interrupt handler functions are functions registered by drivers that
+core shall call when the related interrupt occurs. Structure
+``struct itr_handler``
+references a handler. It contains the handler function entry point as well
+as interrupt information: its number, its interrupt controller device and
+a few other fields.
```
```suggestion
references a handler. It contains the handler function entry point as well
as interrupt information such as the interrupt number, interrupt controller
device, and a few more parameters.
```

etienne-lms replied:
thanks. I've slightly rephrased the information.

jenswi-linaro commented on architecture/core.rst:
```diff
@@ -390,6 +390,186 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+    static void foo_release(void)
+    {
+            if (foo_int1_handler) {
+                    interrupt_disale(foo_int1_handler->chip,
```
interrupt_disable

jbech-linaro commented on architecture/core.rst:
```diff
@@ -390,6 +390,186 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+It shall return ``ITRR_HANDLED`` when the interrupt is served and
+``ITRR_NONE`` when the interrupt cannot be served.
+
+The interrupt handler executes in an interrupt context, not in a thread
+context. This means all other interrupts are masked in the mean time,
+the handler must therefore execute as fast as possible. If there are
```
Suggestion for the sentence starting with "The interrupt" to the next sentence that ends with "... fast as possible".

```suggestion
The interrupt handler runs in an interrupt context rather than a thread
context. When this occurs, all other interrupts are masked, necessitating fast
execution of the interrupt handler to avoid delaying or missing out on other
interrupts.
```

jbech-linaro replied:
Suggestion for the sentence starting with "The interrupt" to the next sentence that ends with "... fast as possible".

```suggestion
The interrupt handler runs in an interrupt context rather than a thread
context. When this occurs, all other interrupts are masked, necessitating fast
execution of the interrupt handler to avoid delaying or missing out on other
interrupts.
```

jbech-linaro commented on architecture/core.rst:
```diff
@@ -390,6 +390,186 @@ normal world supports scheduling of processes. Even on UP systems, supporting
+**Using device tree**
+
+Device drivers can rely on Device Tree to get references to the
+interrupt they consume. Refer to Device Tree document
```
"document" or "documentation"?

etienne-lms replied:
indeed

etienne-lms commented:
I'll remove commit "faq: Fix typo" that is completely unrelated and already addressed by https://github.com/OP-TEE/optee_docs/pull/190. Sorry for the noise.

jbech-linaro commented:
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

jforissier commented:
My ack still holds.

etienne-lms commented:
tags applied.
thanks for the reviews and feedback.

Please remember to not merge this change until https://github.com/OP-TEE/optee_os/pull/5954 (or equiv) is merged.

etienne-lms commented:
Updated according to https://github.com/OP-TEE/optee_os/pull/5954 latest state.

etienne-lms commented:
Added a fixup commit for a typo in the provided example.

Added a section exampling how to use aysnc notif + an interrupt to relax CPU in a thread execution context.

Added a section on a interrupt framework evolution to allows a interrupt consumer to register a bottom half handler function to be called in OP-TEE common bottom half notification thread. There is no P-R created for the changes required in _interrupt.c/.h_, i'll create a P-R later, if the API presented here looks OK.
